File: src/generator/completionGenerator.ts
import { CallbackManager, Traceable } from "../utils/callbacks";
import { CompletionModel } from "./completion-models/completionModel";

/**
 * A CompletionGenerator generates some response resulting from completion requests to
 * one or more CompletionModels.
 */
export abstract class CompletionGenerator<P, R> implements Traceable {
  callbackManager?: CallbackManager;

  constructor(callbackManager?: CallbackManager) {
    this.callbackManager = callbackManager;
  }

  /**
   * Perform completion generation using the given parameters and return response
   * data
   * @param params The parameters to use for generating the completion
   * @returns A promise that resolves to the generated completion data
   */
  abstract run(params: P): Promise<R>;
}

/**
 * LLM Completion Generators are used for generating some completion response from an
 * LLM Completion Model based on input parameters and any applicable internal logic.
 */
export abstract class LLMCompletionGenerator<
    M extends CompletionModel<MP, MR>,
    P,
    R,
    MP = M extends CompletionModel<infer Param, infer _MR> ? Param : never,
    MR = M extends CompletionModel<MP, infer Result> ? Result : never,
  >
  extends CompletionGenerator<P, R>
  implements Traceable
{
  model: M;

  constructor(model: M, callbackManager?: CallbackManager) {
    super(callbackManager);
    this.model = model;
  }
}

File: python/src/semantic_retrieval/generator/completion_generator.py
from abc import abstractmethod
from dataclasses import dataclass
from typing import Generic, Optional

from semantic_retrieval.common.types import P, R
from semantic_retrieval.generator.completion_models.completion_model import (
    CompletionModel,
    CompletionModelParams,
)
from semantic_retrieval.utils.callbacks import CallbackManager, Traceable


@dataclass
class LLMCompletionGeneratorParams(Generic[P], CompletionModelParams[P]):
    pass


class LLMCompletionGenerator(Generic[P, R], Traceable):
    def __init__(
        self,
        model: CompletionModel[P, R],
        callback_manager: Optional[CallbackManager] = None,
    ):
        self.model = model
        self.callback_manager = callback_manager

    @abstractmethod
    async def run(self, params: P) -> R:
        pass
